File: app/src/main/java/de/cineaste/android/database/dbHelper/SeriesDbHelper.java
package de.cineaste.android.database.dbHelper;


import android.content.Context;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import de.cineaste.android.database.dao.BaseDao;
import de.cineaste.android.database.dao.EpisodeDao;
import de.cineaste.android.database.dao.SeasonDao;
import de.cineaste.android.database.dao.SeriesDao;
import de.cineaste.android.entity.series.Episode;
import de.cineaste.android.entity.series.Season;
import de.cineaste.android.entity.series.Series;
import de.cineaste.android.fragment.WatchState;

import static de.cineaste.android.database.dao.BaseDao.EpisodeEntry;
import static de.cineaste.android.database.dao.BaseDao.SeasonEntry;
import static de.cineaste.android.database.dao.BaseDao.SeriesEntry;

public class SeriesDbHelper {
    
    private static SeriesDbHelper instance;

    private final SeriesDao seriesDao;
    private final SeasonDao seasonDao;
    private final EpisodeDao episodeDao;

    private SeriesDbHelper(Context context) {
        this.seriesDao = SeriesDao.getInstance(context);
        this.seasonDao = SeasonDao.getInstance(context);
        this.episodeDao = EpisodeDao.getInstance(context);
    }

    public static SeriesDbHelper getInstance(Context context) {
        if (instance == null) {
            instance = new SeriesDbHelper(context);
        }
        return instance;
    }
    
    public Series getSeriesById(long seriesId) {
        String selection = SeriesEntry._ID + " = ?";
        String[] selectionArgs = {Long.toString(seriesId)};

        List<Series> seriesList = seriesDao.read(selection, selectionArgs, null);

        if (seriesList.size() == 0) {
            return null;
        }

        Series series = seriesList.get(0);
        loadRemainingInformation(series);

        return series;
    }

    public List<Series> getAllSeries() {
        List<Series> seriesList = seriesDao.read(null, null, null);

        for (Series series : seriesList) {
            loadRemainingInformation(series);
        }

        return seriesList;
    }

    public List<Series> getSeriesByWatchedState(WatchState watchedState) {
        String selectionArg = getSelectionArgs(watchedState);
        String selection = SeriesEntry.COLUMN_SERIES_SERIES_WATCHED + " = ?";
        String[] selectionArgs = {selectionArg};

        List<Series> seriesList = seriesDao.read(selection, selectionArgs, SeriesEntry.COLUMN_SERIES_LIST_POSITION + " ASC");

        for (Series series : seriesList) {
            loadRemainingInformation(series);
        }

        return seriesList;
    }

    public List<Episode> getEpisodesBySeasonId(long seasonId) {
        String selection = BaseDao.EpisodeEntry.COLUMN_EPISODE_SEASON_ID + " = ?";
        String[] selectionArgs = {String.valueOf(seasonId)};

        return episodeDao.read(selection, selectionArgs);
    }

    public List<Series> reorderAlphabetical(WatchState state) {
        return reorder(state, SeriesEntry.COLUMN_SERIES_NAME);
    }

    public List<Series> reorderByReleaseDate(WatchState state) {
        return reorder(state, SeriesEntry.COLUMN_SERIES_RELEASE_DATE);
    }

    private List<Series> reorder(WatchState state, String orderBy) {
        String sql = "UPDATE " + SeriesEntry.TABLE_NAME +
                " SET " + SeriesEntry.COLUMN_SERIES_LIST_POSITION + " = ( " +
                "SELECT COUNT(*) " +
                "FROM " + SeriesEntry.TABLE_NAME + " AS t2 " +
                "WHERE t2." + orderBy + " <= " + SeriesEntry.TABLE_NAME + "." + orderBy +
                " ) " +
                "WHERE " + SeriesEntry.COLUMN_SERIES_SERIES_WATCHED + " = " + getSelectionArgs(state);
        seriesDao.reorder(sql);

        return getSeriesByWatchedState(state);
    }
    
    private void loadRemainingInformation(Series series) {
        List<Season> seasons = readSeasonsBySeriesId(series.getId());

        for (Season season : seasons) {
            season.setEpisodes(readEpisodesBySeasonId(season.getId()));
        }
        
        series.setSeasons(seasons);
    }
    
    private List<Season> readSeasonsBySeriesId(long seriesId) {
        String selection = SeasonEntry.COLUMN_SEASON_SERIES_ID + " = ?";
        String[] selectionArgs = {Long.toString(seriesId)};
        
        return seasonDao.read(selection, selectionArgs);
    }
    
    private List<Episode> readEpisodesBySeasonId(long seasonId) {
        String selection = EpisodeEntry.COLUMN_EPISODE_SEASON_ID + " = ?";
        String[] selectionArgs = {String.valueOf(seasonId)};

        return episodeDao.read(selection, selectionArgs);
    }
    
    public void addToWatchList(Series series) {
        addToList(series, false);
    }

    public void addToHistory(Series series) {
        addToList(series, true);
    }

    private void addToList(Series series, boolean watchState) {
        series.setWatched(watchState);
        series.setListPosition(seriesDao.getHighestListPosition(watchState));
        seriesDao.create(series);
        for (Season season : series.getSeasons()) {
            season.setWatched(watchState);
            seasonDao.create(season, series.getId());
            for (Episode episode : season.getEpisodes()) {
                episode.setWatched(watchState);
                episodeDao.create(episode, series.getId(), season.getId());
            }
        }
    }

    public void moveToWatchList(Series series) {
        moveBetweenLists(series, false);
    }

    public void moveToHistory(Series series) {
        moveBetweenLists(series, true);
    }

    private void moveBetweenLists(Series series, boolean watchState) {
        String updateEpisodesSql = "UPDATE " + EpisodeEntry.TABLE_NAME +
                " SET " + EpisodeEntry.COLUMN_EPISODE_WATCHED + " = " + String.valueOf(watchState ? 1 : 0) +
                " WHERE " + EpisodeEntry.COLUMN_EPISODE_SERIES_ID + " = " + series.getId();

        String updateSeasonsSql = "UPDATE " + SeasonEntry.TABLE_NAME +
                " SET " + SeasonEntry.COLUMN_SEASON_WATCHED + " = " + String.valueOf(watchState ? 1 : 0) +
                " WHERE " + SeasonEntry.COLUMN_SEASON_SERIES_ID + " = " + series.getId();

        series.setWatched(watchState);

        seriesDao.update(series, seriesDao.getHighestListPosition(watchState));
        seasonDao.executeCustomSql(updateSeasonsSql);
        episodeDao.executeCustomSql(updateEpisodesSql);
    }

    public void moveBackToWatchList(Series series, int prevSeason, int prevEpisode) {
        moveBackToList(series, prevSeason, prevEpisode, false);
    }

    public void moveBackToHistory(Series series, int prevSeason, int prevEpisode) {
        moveBackToList(series, prevSeason, prevEpisode, true);
    }

    private void moveBackToList(Series series, int prevSeason, int prevEpisode, boolean watchState) {
        moveBetweenLists(series, watchState);
        for (Season season : series.getSeasons()) {
            if (season.getSeasonNumber() < prevSeason) {
                season.setWatched(!watchState);
                seasonDao.update(season);

                String updateEpisodesSql = "UPDATE " + EpisodeEntry.TABLE_NAME +
                        " SET " + EpisodeEntry.COLUMN_EPISODE_WATCHED + " = " + String.valueOf(!watchState ? 1 : 0) +
                        " WHERE " + EpisodeEntry.COLUMN_EPISODE_SEASON_ID + " = " + season.getId();

                episodeDao.executeCustomSql(updateEpisodesSql);
            }

            if (season.getSeasonNumber() == prevSeason) {
                for (Episode episode : season.getEpisodes()) {
                    if (episode.getEpisodeNumber() < prevEpisode) {
                        episode.setWatched(!watchState);
                        episodeDao.update(episode);
                    }
                }
            }
        }
    }

    public void delete(Series series) {
        episodeDao.deleteBySeriesId(series.getId());
        seasonDao.deleteBySeriesId(series.getId());
        seriesDao.delete(series.getId());
    }

    public void delete(long seriesId) {
        episodeDao.deleteBySeriesId(seriesId);
        seasonDao.deleteBySeriesId(seriesId);
        seriesDao.delete(seriesId);
    }
    
    public void episodeWatched(Series series) {
        List<Season> seasons = new ArrayList<>();
        for (Season season : series.getSeasons()) {
            if (season.isWatched()) {
                seasons.add(season);
            } else {
                List<Episode> episodes = new ArrayList<>();
                for (Episode episode : season.getEpisodes()) {
                    if (episode.isWatched()) {
                        episodes.add(episode);
                    } else {
                        episode.setWatched(true);
                        episodeDao.update(episode);
                        episodes.add(episode);
                        break;
                    }
                }
                if (episodes.size() == season.getEpisodes().size()) {
                    season.setWatched(true);
                    seasonDao.update(season);
                    seasons.add(season);
                }
                break;
            }
        }

        if (seasons.size() == series.getSeasons().size() && !series.isInProduction()) {
            series.setWatched(true);
            seriesDao.update(series, seriesDao.getHighestListPosition(true));
        }
    }

    public void episodeClicked(Episode episode) {
        episode.setWatched(!episode.isWatched());
        episodeDao.update(episode);

        if (!episode.isWatched()) {
            Series series = getSeriesById(episode.getSeriesId());
            series.setWatched(false);
            seriesDao.update(series, seriesDao.getHighestListPosition(false));
            for (Season season : series.getSeasons()) {
                if (season.getId() == episode.getSeasonId()) {
                    season.setWatched(false);
                    seasonDao.update(season);
                    break;
                }
            }

        }
    }

    public void importSeries(Series series) {
        Series oldSeries = getSeriesById(series.getId());
        if (oldSeries == null) {
            seriesDao.create(series);
            for (Season season : series.getSeasons()) {
                seasonDao.create(season, series.getId());
                for (Episode episode : season.getEpisodes()) {
                    episodeDao.create(episode, series.getId(), season.getId());
                }
            }
        } else {
            seriesDao.update(series, series.getListPosition());
            for (Season season : series.getSeasons()) {
                seasonDao.update(season);
                for (Episode episode : season.getEpisodes()) {
                    episodeDao.update(episode);
                }
            }
        }
    }
    
    public void update(Series series) {
        Series oldSeries = getSeriesById(series.getId());
        int newPosition = getNewPosition(series, oldSeries);
        series.setWatched(oldSeries.isWatched());

        seriesDao.update(series, newPosition);

        for (Season season : series.getSeasons()) {
            update(season, series.getId());
        }

    }

    public void updatePosition(Series series) {
        Series oldSeries = getSeriesById(series.getId());
        int newPosition = getNewPosition(series, oldSeries);
        seriesDao.update(series, newPosition);
    }

    private void update(Season season, long seriesId) {
        String selection = SeasonEntry._ID + " = ?";
        String[] selectionArgs = {Long.toString(season.getId())};

        season.setSeriesId(seriesId);

        List<Season> seasons = seasonDao.read(selection, selectionArgs);
        if (seasons.size() > 0) {
            Season oldSeason = seasonDao.read(selection, selectionArgs).get(0);
            season.setWatched(oldSeason.isWatched());
            seasonDao.update(season);
        } else {
            seasonDao.create(season, seriesId);
        }


        for (Episode episode : season.getEpisodes()) {
            update(episode, seriesId, season.getId());
        }
    }

    private void update(Episode episode, long seriesId, long seasonId) {
        String selection = EpisodeEntry._ID + " = ?";
        String[] selectionArgs = {Long.toString(episode.getId())};

        episode.setSeriesId(seriesId);
        episode.setSeasonId(seasonId);

        List<Episode> episodes = episodeDao.read(selection, selectionArgs);

        if (episodes.size() > 0) {
            Episode oldEpisode = episodes.get(0);
            episode.setWatched(oldEpisode.isWatched());
            episodeDao.update(episode);
        } else {
            episodeDao.create(episode);
        }

    }

    private int getNewPosition(Series updatedSeries, Series oldSeries) {
        if (updatedSeries.isWatched() == oldSeries.isWatched()) {
            return updatedSeries.getListPosition();
        }

        return seriesDao.getHighestListPosition(updatedSeries.isWatched());
    }

    @NonNull
    private String getSelectionArgs(WatchState state) {
        String selectionArg;
        if (state == WatchState.WATCH_STATE) {
            selectionArg = "0";
        } else {
            selectionArg = "1";
        }
        return selectionArg;
    }
}

File: app/src/main/kotlin/de/cineaste/android/database/dbHelper/SeriesDbHelper.kt
package de.cineaste.android.database.dbHelper

import android.content.Context
import java.util.ArrayList
import de.cineaste.android.database.dao.BaseDao.EpisodeEntry
import de.cineaste.android.database.dao.BaseDao.SeasonEntry
import de.cineaste.android.database.dao.BaseDao.SeriesEntry
import de.cineaste.android.database.dao.EpisodeDao
import de.cineaste.android.database.dao.SeasonDao
import de.cineaste.android.database.dao.SeriesDao
import de.cineaste.android.entity.series.Episode
import de.cineaste.android.entity.series.Season
import de.cineaste.android.entity.series.Series
import de.cineaste.android.fragment.WatchState

class SeriesDbHelper private constructor(context: Context) {

    private val seriesDao: SeriesDao = SeriesDao.getInstance(context)
    private val seasonDao: SeasonDao = SeasonDao.getInstance(context)
    private val episodeDao: EpisodeDao = EpisodeDao.getInstance(context)

    val allSeries: List<Series>
        get() {
            val seriesList = seriesDao.read(null, null, null)

            for (series in seriesList) {
                loadRemainingInformation(series)
            }

            return seriesList
        }

    fun getSeriesById(seriesId: Long): Series? {
        val selection = SeriesEntry.ID + " = ?"
        val selectionArgs = arrayOf(seriesId.toString())

        val seriesList = seriesDao.read(selection, selectionArgs, null)

        if (seriesList.isEmpty()) {
            return null
        }

        val series = seriesList[0]
        loadRemainingInformation(series)

        return series
    }

    fun getSeriesByWatchedState(watchedState: WatchState): List<Series> {
        val selectionArg = getSelectionArgs(watchedState)
        val selection = SeriesEntry.COLUMN_SERIES_SERIES_WATCHED + " = ?"
        val selectionArgs = arrayOf(selectionArg)

        val seriesList = seriesDao.read(
            selection,
            selectionArgs,
            SeriesEntry.COLUMN_SERIES_LIST_POSITION + " ASC"
        )

        for (series in seriesList) {
            loadRemainingInformation(series)
        }

        return seriesList
    }

    fun getUnWatchedEpisodesOfSeries(seriesId: Long): List<Episode> {
        val selection =
            EpisodeEntry.COLUMN_EPISODE_SERIES_ID + " = ? AND " + EpisodeEntry.COLUMN_EPISODE_WATCHED + " = 0"
        val selectionArgs = arrayOf(seriesId.toString())

        return episodeDao.read(selection, selectionArgs)
    }

    fun getEpisodesBySeasonId(seasonId: Long): List<Episode> {
        val selection = EpisodeEntry.COLUMN_EPISODE_SEASON_ID + " = ?"
        val selectionArgs = arrayOf(seasonId.toString())

        return episodeDao.read(selection, selectionArgs)
    }

    fun reorderAlphabetical(state: WatchState): List<Series> {
        return reorder(state, SeriesEntry.COLUMN_SERIES_NAME)
    }

    fun reorderByReleaseDate(state: WatchState): List<Series> {
        return reorder(state, SeriesEntry.COLUMN_SERIES_RELEASE_DATE)
    }

    private fun reorder(state: WatchState, orderBy: String): List<Series> {
        val sql = "UPDATE " + SeriesEntry.TABLE_NAME +
                " SET " + SeriesEntry.COLUMN_SERIES_LIST_POSITION + " = ( " +
                "SELECT COUNT(*) " +
                "FROM " + SeriesEntry.TABLE_NAME + " AS t2 " +
                "WHERE t2." + orderBy + " <= " + SeriesEntry.TABLE_NAME + "." + orderBy +
                " ) " +
                "WHERE " + SeriesEntry.COLUMN_SERIES_SERIES_WATCHED + " = " + getSelectionArgs(state)
        seriesDao.reorder(sql)

        return getSeriesByWatchedState(state)
    }

    private fun loadRemainingInformation(series: Series) {
        val seasons = readSeasonsBySeriesId(series.id)

        for (season in seasons) {
            season.episodes = readEpisodesBySeasonId(season.id)
        }

        series.seasons = seasons
    }

    private fun getSeasonById(seasonId: Long): Season? {
        val selection = SeasonEntry.ID + " = ?"
        val selectionArgs = arrayOf(seasonId.toString())

        val seasonList = seasonDao.read(selection, selectionArgs)

        return if (seasonList.isEmpty()) null else seasonList[0]
    }

    private fun readSeasonsBySeriesId(seriesId: Long): List<Season> {
        val selection = SeasonEntry.COLUMN_SEASON_SERIES_ID + " = ?"
        val selectionArgs = arrayOf(seriesId.toString())

        return seasonDao.read(selection, selectionArgs)
    }

    private fun readEpisodesBySeasonId(seasonId: Long): List<Episode> {
        val selection = EpisodeEntry.COLUMN_EPISODE_SEASON_ID + " = ?"
        val selectionArgs = arrayOf(seasonId.toString())

        return episodeDao.read(selection, selectionArgs)
    }

    fun addToWatchList(series: Series) {
        addToList(series, false)
    }

    fun addToHistory(series: Series) {
        addToList(series, true)
    }

    private fun addToList(series: Series, watchState: Boolean) {
        series.isWatched = watchState
        series.listPosition = seriesDao.getHighestListPosition(watchState)
        seriesDao.create(series)
        for (season in series.seasons) {
            season.isWatched = watchState
            seasonDao.create(season, series.id)
            for (episode in season.episodes) {
                episode.isWatched = watchState
                episodeDao.create(episode, series.id, season.id)
            }
        }
    }

    fun moveToWatchList(series: Series) {
        moveBetweenLists(series, false)
    }

    fun moveToHistory(series: Series) {
        moveBetweenLists(series, true)
    }

    fun toggleSeason(seasonId: Long) {
        getSeasonById(seasonId)?.let { season ->
            val watchState = !season.isWatched
            val updateEpisodesSql = "UPDATE " + EpisodeEntry.TABLE_NAME +
                    " SET " + EpisodeEntry.COLUMN_EPISODE_WATCHED + " = " + (if (watchState) 1 else 0).toString() +
                    " WHERE " + EpisodeEntry.COLUMN_EPISODE_SEASON_ID + " = " + seasonId

            val updateSeasonsSql = "UPDATE " + SeasonEntry.TABLE_NAME +
                    " SET " + SeasonEntry.COLUMN_SEASON_WATCHED + " = " + (if (watchState) 1 else 0).toString() +
                    " WHERE " + SeasonEntry.ID + " = " + seasonId

            seasonDao.executeCustomSql(updateSeasonsSql)
            episodeDao.executeCustomSql(updateEpisodesSql)
        }
    }

    private fun moveBetweenLists(series: Series, watchState: Boolean) {
        val updateEpisodesSql = "UPDATE " + EpisodeEntry.TABLE_NAME +
                " SET " + EpisodeEntry.COLUMN_EPISODE_WATCHED + " = " + (if (watchState) 1 else 0).toString() +
                " WHERE " + EpisodeEntry.COLUMN_EPISODE_SERIES_ID + " = " + series.id

        val updateSeasonsSql = "UPDATE " + SeasonEntry.TABLE_NAME +
                " SET " + SeasonEntry.COLUMN_SEASON_WATCHED + " = " + (if (watchState) 1 else 0).toString() +
                " WHERE " + SeasonEntry.COLUMN_SEASON_SERIES_ID + " = " + series.id

        series.isWatched = watchState

        seriesDao.update(series, seriesDao.getHighestListPosition(watchState))
        seasonDao.executeCustomSql(updateSeasonsSql)
        episodeDao.executeCustomSql(updateEpisodesSql)
    }

    fun moveBackToWatchList(series: Series, prevSeason: Int, prevEpisode: Int) {
        moveBackToList(series, prevSeason, prevEpisode, false)
    }

    fun moveBackToHistory(series: Series, prevSeason: Int, prevEpisode: Int) {
        moveBackToList(series, prevSeason, prevEpisode, true)
    }

    private fun moveBackToList(
        series: Series,
        prevSeason: Int,
        prevEpisode: Int,
        watchState: Boolean
    ) {
        moveBetweenLists(series, watchState)
        for (season in series.seasons) {
            if (season.seasonNumber < prevSeason) {
                season.isWatched = !watchState
                seasonDao.update(season)

                val updateEpisodesSql = "UPDATE " + EpisodeEntry.TABLE_NAME +
                        " SET " + EpisodeEntry.COLUMN_EPISODE_WATCHED + " = " + (if (!watchState) 1 else 0).toString() +
                        " WHERE " + EpisodeEntry.COLUMN_EPISODE_SEASON_ID + " = " + season.id

                episodeDao.executeCustomSql(updateEpisodesSql)
            }

            if (season.seasonNumber == prevSeason) {
                for (episode in season.episodes) {
                    if (episode.episodeNumber < prevEpisode) {
                        episode.isWatched = !watchState
                        episodeDao.update(episode)
                    }
                }
            }
        }
    }

    fun delete(series: Series) {
        episodeDao.deleteBySeriesId(series.id)
        seasonDao.deleteBySeriesId(series.id)
        seriesDao.delete(series.id)
    }

    fun delete(seriesId: Long) {
        episodeDao.deleteBySeriesId(seriesId)
        seasonDao.deleteBySeriesId(seriesId)
        seriesDao.delete(seriesId)
    }

    fun episodeWatched(series: Series) {
        val seasons = ArrayList<Season>()
        for (season in series.seasons) {
            if (season.isWatched) {
                seasons.add(season)
            } else {
                val episodes = ArrayList<Episode>()
                for (episode in season.episodes) {
                    if (episode.isWatched) {
                        episodes.add(episode)
                    } else {
                        episode.isWatched = true
                        episodeDao.update(episode)
                        episodes.add(episode)
                        break
                    }
                }
                if (episodes.size == season.episodes.size) {
                    season.isWatched = true
                    seasonDao.update(season)
                    seasons.add(season)
                }
                break
            }
        }

        if (seasons.size == series.seasons.size && !series.isInProduction) {
            series.isWatched = true
            seriesDao.update(series, seriesDao.getHighestListPosition(true))
        }
    }

    fun episodeClicked(episode: Episode) {
        episode.isWatched = !episode.isWatched
        episodeDao.update(episode)

        if (!episode.isWatched) {
            val series = getSeriesById(episode.seriesId)
            series?.let {
                series.isWatched = false
                seriesDao.update(series, seriesDao.getHighestListPosition(false))
                for (season in series.seasons) {
                    if (season.id == episode.seasonId) {
                        season.isWatched = false
                        seasonDao.update(season)
                        break
                    }
                }
            }
        }
    }

    fun importSeries(series: Series) {
        val oldSeries = getSeriesById(series.id)
        if (oldSeries == null) {
            seriesDao.create(series)
            for (season in series.seasons) {
                seasonDao.create(season, series.id)
                for (episode in season.episodes) {
                    episodeDao.create(episode, series.id, season.id)
                }
            }
        } else {
            seriesDao.update(series, series.listPosition)
            for (season in series.seasons) {
                seasonDao.update(season)
                for (episode in season.episodes) {
                    episodeDao.update(episode)
                }
            }
        }
    }

    fun updateWatchState(series: Series) {
        val newListPosition = seriesDao.getHighestListPosition(series.isWatched)
        series.listPosition = newListPosition

        seriesDao.update(series, newListPosition)
    }

    fun update(series: Series) {
        getSeriesById(series.id)?.let { oldSeries ->
            val newPosition = getNewPosition(series, oldSeries)
            series.isWatched = oldSeries.isWatched
            series.listPosition = newPosition

            seriesDao.update(series, newPosition)

            for (season in series.seasons) {
                update(season, series)
            }
        }
    }

    fun updatePosition(series: Series) {
        getSeriesById(series.id)?.let { oldSeries ->
            val newPosition = getNewPosition(series, oldSeries)
            seriesDao.update(series, newPosition)
        }
    }

    private fun update(season: Season, series: Series) {
        val selection = SeasonEntry.ID + " = ?"
        val selectionArgs = arrayOf(java.lang.Long.toString(season.id))

        season.seriesId = series.id

        val seasons = seasonDao.read(selection, selectionArgs)
        if (seasons.isNotEmpty()) {
            val oldSeason = seasonDao.read(selection, selectionArgs)[0]
            season.isWatched = oldSeason.isWatched
            seasonDao.update(season)
        } else {
            season.isWatched = false
            seasonDao.create(season, series.id)

            series.isWatched = false
            seriesDao.update(series, series.listPosition)
        }

        for (episode in season.episodes) {
            update(episode, series, season)
        }
    }

    private fun update(episode: Episode, series: Series, season: Season) {
        val selection = EpisodeEntry.ID + " = ?"
        val selectionArgs = arrayOf(java.lang.Long.toString(episode.id))

        episode.seriesId = series.id
        episode.seasonId = season.id

        val episodes = episodeDao.read(selection, selectionArgs)

        if (episodes.isNotEmpty()) {
            val oldEpisode = episodes[0]
            episode.isWatched = oldEpisode.isWatched
            episodeDao.update(episode)
        } else {
            episode.isWatched = false
            episodeDao.create(episode)

            season.isWatched = false
            seasonDao.update(season)

            series.isWatched = false
            seriesDao.update(series, series.listPosition)
        }
    }

    private fun getNewPosition(updatedSeries: Series, oldSeries: Series): Int {
        return if (updatedSeries.isWatched == oldSeries.isWatched) {
            updatedSeries.listPosition
        } else seriesDao.getHighestListPosition(updatedSeries.isWatched)
    }

    private fun getSelectionArgs(state: WatchState): String {
        return if (state == WatchState.WATCH_STATE) "0" else "1"
    }

    companion object {

        private var instance: SeriesDbHelper? = null

        fun getInstance(context: Context): SeriesDbHelper {
            if (instance == null) {
                instance = SeriesDbHelper(context)
            }
            return instance!!
        }
    }
}
